controllers/apps/transformer_component_deletion.go

```diff
@@ -36,42 +54,126 @@ func (t *componentDeletionTransformer) Transform(ctx graph.TransformContext, dag
 		return nil
 	}
 
+	reqCtx := intctrlutil.RequestCtx{
+		Ctx:      transCtx.Context,
+		Log:      transCtx.Logger,
+		Recorder: transCtx.EventRecorder,
+	}
 	graphCli, _ := transCtx.Client.(model.GraphClient)
 	comp := transCtx.Component
+	cluster, err := t.getCluster(transCtx, comp)
+	if err != nil {
+		return newRequeueError(requeueDuration, err.Error())
+	}
 
-	// clusterName, err := getClusterName(comp)
-	// if err != nil {
-	//	return newRequeueError(requeueDuration, err.Error())
-	// }
-	//
-	// compName, err := component.ShortName(clusterName, comp.Name)
-	// ml := constant.GetComponentWellKnownLabels(clusterName, compName)
-	// snapshot, err := model.ReadCacheSnapshot(transCtx, comp, ml, kindsForCompDelete()...)
-	// if err != nil {
-	//	return newRequeueError(requeueDuration, err.Error())
-	// }
-	// for _, object := range snapshot {
-	//	graphCli.Delete(dag, object)
-	// }
-
-	comp.Status.Phase = appsv1alpha1.DeletingClusterCompPhase
-	graphCli.Delete(dag, comp)
+	// step1: update the component status to deleting
+	if comp.Status.Phase != appsv1alpha1.DeletingClusterCompPhase {
+		comp.Status.Phase = appsv1alpha1.DeletingClusterCompPhase
+		graphCli.Status(dag, comp, transCtx.Component)
+		return newRequeueError(time.Second*1, "updating component status to deleting")
+	}
+
+	// step2: do the pre-terminate action if needed
+	if err := component.ReconcileCompPreTerminate(reqCtx, t.Client, cluster, comp, dag); err != nil {
+		if intctrlutil.IsTargetError(err, intctrlutil.ErrorWaitCacheRefresh) {
```

Comment:
what's the ErrorWaitCacheRefresh error? and is there anything special about it here?

Reply:
> what's the ErrorWaitCacheRefresh error? and is there anything special about it here?

 updated.  ErrorWaitCacheRefresh is primarily used to handle some anticipated time-consuming waiting tasks. In this case, it is repurposed to wait for the completion of a job execution, avoiding unnecessary reconciliations. Once the job is finished, a new reconciliation will be triggered through an event.

Reply:
> > what's the ErrorWaitCacheRefresh error? and is there anything special about it here?
> 
> updated. ErrorWaitCacheRefresh is primarily used to handle some anticipated time-consuming waiting tasks. In this case, it is repurposed to wait for the completion of a job execution, avoiding unnecessary reconciliations. Once the job is finished, a new reconciliation will be triggered through an event.

ErrorWaitCacheRefresh -> ErrorTypeExpectedInProcess instead